amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionSettings.scala

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.amqp
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import com.rabbitmq.client.{Address, Connection, ConnectionFactory, ExceptionHandler}
+
+/**
+ * Only for internal implementations
+ */
+sealed trait AmqpConnectionSettings {
+  def getConnection: Connection
+}
+
+final case class ReusableAmqpConnectionSettings(settings: AmqpConnectionSettings) extends AmqpConnectionSettings {
+  private var cachedConnection: Option[Connection] = None
+
+  override def getConnection = cachedConnection match {
+    case Some(connection) => {
+      if (connection.isOpen) connection
+      else {
+        val connection = settings.getConnection
+        cachedConnection = Some(connection)
+        connection
+      }
+    }
+    case None => {
+      val connection = settings.getConnection
+      cachedConnection = Some(connection)
+      connection
+    }
+  }
+
+  def releaseConnection() = cachedConnection match {
+    case Some(connection) => {
+      if (connection.isOpen) connection.close()
+      cachedConnection = None
+    }
+    case None => Unit
+  }
+}
+
+final case class ReusableAmqpConnectionSettingsWithAutomaticRelease(settings: AmqpConnectionSettings)
```

Comment:
And then make this a `CachedConnectionProvider`

Reply:
The previous one is also cached. the only different is that this one release the connection when the last client disconnects and then open a new one if you reuse it again.